hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/freon/TestOMSnapshotDAG.java

```diff
@@ -0,0 +1,152 @@
+  public static void init() throws Exception {
+    conf = new OzoneConfiguration();
+    DatanodeRatisServerConfig ratisServerConfig =
+        conf.getObject(DatanodeRatisServerConfig.class);
+    ratisServerConfig.setRequestTimeOut(Duration.ofSeconds(3));
+    ratisServerConfig.setWatchTimeOut(Duration.ofSeconds(3));
+    conf.setFromObject(ratisServerConfig);
+
+    RatisClientConfig.RaftConfig raftClientConfig =
+        conf.getObject(RatisClientConfig.RaftConfig.class);
+    raftClientConfig.setRpcRequestTimeout(Duration.ofSeconds(3));
+    raftClientConfig.setRpcWatchRequestTimeout(Duration.ofSeconds(3));
+    conf.setFromObject(raftClientConfig);
+
+    cluster = MiniOzoneCluster.newBuilder(conf).setNumDatanodes(5).build();
+    cluster.waitForClusterToBeReady();
+
+    store = cluster.getClient().getObjectStore();
+
+    GenericTestUtils.setLogLevel(RaftLog.LOG, Level.INFO);
+    GenericTestUtils.setLogLevel(RaftServer.LOG, Level.INFO);
+  }
+
+  /**
+   * Shutdown MiniDFSCluster.
+   */
+  @AfterAll
+  public static void shutdown() {
+    if (cluster != null) {
+      LOG.warn("Waiting for an extra 10 seconds before shutting down...");
+      try {
+        Thread.sleep(10000);
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+      cluster.shutdown();
+    }
+  }
+
+  @Test
+  void testZeroSizeKey() throws IOException {
+
+    RandomKeyGenerator randomKeyGenerator =
+        new RandomKeyGenerator(cluster.getConf());
+    CommandLine cmd = new CommandLine(randomKeyGenerator);
+    cmd.execute("--num-of-volumes", "1",
+        "--num-of-buckets", "1",
+        "--num-of-keys", "6000",
+        "--num-of-threads", "1",
+        "--key-size", "0",
+        "--factor", "THREE",
+        "--type", "RATIS",
+        "--validate-writes"
+    );
+
+    Assert.assertEquals(6000L, randomKeyGenerator.getNumberOfKeysAdded());
+    Assert.assertEquals(6000L,
+        randomKeyGenerator.getSuccessfulValidationCount());
+
+    List<OmVolumeArgs> volList = cluster.getOzoneManager()
+        .listAllVolumes("", "", 10);
+    System.out.println(volList);
+    final String volumeName = volList.stream().filter(e ->
+        !e.getVolume().equals(OZONE_S3_VOLUME_NAME_DEFAULT))  // Ignore s3v vol
+        .collect(Collectors.toList()).get(0).getVolume();
+    List<OmBucketInfo> bucketList =
+        cluster.getOzoneManager().listBuckets(volumeName, "", "", 10);
+    System.out.println(bucketList);
+    final String bucketName = bucketList.get(0).getBucketName();
+
+    // Create snapshot
+    String resp = store.createSnapshot(volumeName, bucketName, "snap1");
+    System.out.println(resp);
+
+    final OzoneVolume volume = store.getVolume(volumeName);
+    final OzoneBucket bucket = volume.getBucket(bucketName);
+
+    for (int i = 0; i < 6000; i++) {
+      bucket.createKey("b_" + i, 0).close();
+    }
+
+    resp = store.createSnapshot(volumeName, bucketName, "snap3");
+    System.out.println(resp);
+  }
+
```

Comment:
We are not really testing the DAG and DAG persistence here.
We could do several things e.g. : 
1) Create Snap1 & Snap2 and check that the DAG correctly returns the diff file.
2) We can also check the DAG persistent part and validating that the  DAG reconstructed after system restart returns the same result as 1) above.

Reply:
Sounds good. I would do this in a follow-up jira together with some new debug options in RDBStore.